File: Sources/AINotebookCore/AutoSaveController.swift
import Foundation

@MainActor
public final class AutoSaveController: ObservableObject {
    public enum Status: Equatable, Sendable {
        case saved
        case unsaved
        case saving
        case error(String)
    }

    @Published public private(set) var status: Status = .saved

    private let debounceMillis: Int
    private let save: @Sendable (String) -> Void
    private var pendingBody: String?
    private var debounceTask: Task<Void, Never>?

    public init(
        debounceMillis: Int = 2_000,
        save: @escaping @Sendable (String) -> Void
    ) {
        self.debounceMillis = debounceMillis
        self.save = save
    }

    public func noteDidChange(_ markdown: String) {
        pendingBody = markdown
        status = .unsaved
        debounceTask?.cancel()
        let delay = UInt64(debounceMillis) * 1_000_000
        debounceTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: delay)
            if Task.isCancelled { return }
            await self?.flush()
        }
    }

    public func manualSave() {
        debounceTask?.cancel()
        Task { [weak self] in await self?.flush() }
    }

    private func flush() async {
        guard let body = pendingBody else { return }
        status = .saving
        save(body)
        pendingBody = nil
        status = .saved
    }
}

File: windows/src/AINotebook.App/Editor/AutoSaveController.cs
using System;
using CommunityToolkit.Mvvm.ComponentModel;
using Microsoft.UI.Dispatching;

namespace AINotebook.App.Editor;

public enum SaveState { Saved, Unsaved, Saving, Error }

public partial class AutoSaveController : ObservableObject
{
    private readonly Action<string> _save;
    private readonly DispatcherQueueTimer _timer;
    private string? _pendingBody;

    [ObservableProperty] public partial SaveState Status { get; private set; } = SaveState.Saved;
    [ObservableProperty] public partial string? ErrorText { get; private set; }

    public AutoSaveController(DispatcherQueue dispatcher, Action<string> save, int debounceMillis = 2000)
    {
        _save = save;
        _timer = dispatcher.CreateTimer();
        _timer.Interval = TimeSpan.FromMilliseconds(debounceMillis);
        _timer.IsRepeating = false;
        _timer.Tick += (_, _) => Flush();
    }

    public void NoteDidChange(string markdown)
    {
        _pendingBody = markdown;       // last-write-wins
        Status = SaveState.Unsaved;
        _timer.Stop();
        _timer.Start();
    }

    public void ManualSave()
    {
        _timer.Stop();
        Flush();
    }

    private void Flush()
    {
        if (_pendingBody is not { } body) return;
        Status = SaveState.Saving;
        try
        {
            _save(body);
            _pendingBody = null;
            Status = SaveState.Saved;
            ErrorText = null;
        }
        catch (Exception ex)
        {
            ErrorText = ex.Message;
            Status = SaveState.Error;
        }
    }

    public bool HasUnsavedChanges => Status is SaveState.Unsaved or SaveState.Saving;
}
